File: js-template/leetcode/editor/cn/merge-k-sorted-lists.js
/*
 * @lc app=leetcode.cn id=23 lang=javascript
 * @lcpr version=30404
 *
 * [23] 合并 K 个升序链表
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    if (lists.length === 0) {
        return null;
    } else if (lists.length === 1) {
        return lists[0];
    } else if (lists.length === 2) {
        return merge(lists[0], lists[1]);
    } else {
        var midIndex = Math.floor(lists.length / 2);
        var one = mergeKLists(lists.slice(midIndex));
        var two = mergeKLists(lists.slice(0, midIndex));
        return merge(one, two);
    }
};

/**
 * @param {ListNode} headA 
 * @param {ListNode} headB 
 */
var merge = function(headA, headB) {
    var head = new ListNode(0, null);
    var cur = head;
    while (headA != null && headB != null) {
        if (headA.val < headB.val) {
            cur.next = headA;
            headA = headA.next;
        } else {
            cur.next = headB;
            headB = headB.next;
        }
        cur = cur.next;
    }
    cur.next = headA || headB;
    return head.next;
}
// @lc code=end

// your test code here


mergeKLists([ListNode.createHead([1,4,5]), ListNode.createHead([1,3,4]), ListNode.createHead([2,6])])


/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 * @lcpr version=30404
 *
 * [23] 合并 K 个升序链表
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n = lists.size();
        ListNode *first = nullptr;
        for (auto &&item : lists) {
            ListNode dummyNode(0), *p = &dummyNode;
            ListNode *second = item;
            while (first && second)
            {
                if (first->val > second->val) {
                    p->next = second;
                    second = second->next;
                } else {
                    p->next = first;
                    first = first->next;
                }
                p = p->next;
            }
            p->next = first ? first : second;
            first = dummyNode.next;
        }
        return first;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    vector<ListNode *> lists = {
        ListNode::createHead({1,4,5}),
        ListNode::createHead({1,3,4}),
        ListNode::createHead({2,6})
    };
    solution.mergeKLists(lists);
}



/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

 */

